### DISCUSSION TITLE
Passing join handle to select! in a loop

### DISCUSSION THREAD

**Discussion by sandeepmukh** [tokio-rs/tokio#4019] [Q&A]
I have some async code (a state machine running an algorithm) running in parallel to processing those messages and then terminating once the state machine returns its final state. Is there a way to do this?

Code outline:
```
let join_handle = tokio::spawn(async move { match state_machine.execute().await  });
loop {
    tokio::select! {
        result = join_handle => { // this is the trouble line!
                  // process result
                  break;
          },
           event = handle_incoming_messages() => { // handle messages },
        }
 }
```

Error:
use of moved value: `join_handle`

value moved here, in previous iteration of looprustc(E0382)
main.rs(169, 9): move occurs because `join_handle` has type `tokio::task::JoinHandle<_>`, which does not implement the `Copy` trait
main.rs(178, 22): value moved here, in previous iteration of loop

**Comment by Darksonn** [comment]
You can do this by passing a mutable reference to the `JoinHandle`.
```Rust
tokio::select! {
    result = &mut join_handle => {
        // process result
        break;
    },
    event = handle_incoming_messages() => {
        // handle messages
    },
}
```
This ensures that the `JoinHandle` is not consumed. Note that it isn't even consumed when the `JoinHandle` completes. Any use of the `JoinHandle` after completion will result in a panic.

**Reply by sandeepmukh** [reply]
Thank you! Do you know how I can check when the join handle completes? I'm trying to use `now_or_never` but it always seems to return None.

**Reply by Darksonn** [reply]
You can do it with `now_or_never`. You can open a new question if you're running into trouble with it.
